sort stories by kind

hipstersmoothie opened the pull request:
Trying to address my comments in #63. I plan to add the `byKind` option to write the different kinds to different files. 

So far my changes have changed what the output looks like for the default mode (single sketch file). Insteaad of all the stories going down vertically, the stories are now going horizontally by kind. I find this a way easier way to consume the output of story2sketch, as I can parse what I'm looking at easier

Here is what it currently looks like. Each column is a different kind (ex: Button, Choice) and the rows in the column are the individual stories.

@chrisvxd what do you think?

<img width="1069" alt="Screen Shot 2019-03-30 at 4 15 37 PM" src="https://user-images.githubusercontent.com/1192452/55282721-9a4b1880-5307-11e9-89cd-1a41cc861a4e.png">

EDIT: separate files are now implemented.

![Screen Shot 2019-03-30 at 4 55 21 PM](https://user-images.githubusercontent.com/1192452/55282964-a4bbe100-530c-11e9-9834-e897fb06f91d.png)

chrisvxd commented on src/server/Story2sketch.js:
```diff
@@ -254,22 +259,55 @@ export default class Story2sketch {
+        xOffset += this.widestByViewport[id] + this.symbolGutter;
+      }
+    }
+  }
```
The changes here make this function incompatible with the original layout. Since I'm keen to retain the original layout, we could:

1. Make this functional, so it accepts an object called `symbolsByViewport` (which is actually keyed on viewport - see other comment regarding adding `symbolsByKind` variable), and returns an array of symbols.

```js
positionSymbols(symbolsByViewport) {
    let layers = [];
    let xOffset = 0;

    for (const { id } of this.sortedViewports) {
      let yOffset = 0;

      const symbols = symbolsByViewport[id];

      for (const [index, symbol] of symbols.entries()) {
        // Skip failed symbols
        if (symbol) {
          symbol.frame.x = xOffset;
          symbol.frame.y = yOffset;
          layers.push(symbol);
          yOffset += this.tallestByStory[index] + this.symbolGutter;
        }
      }

      xOffset += this.widestByViewport[id] + this.symbolGutter;
    }
    
    return layers;
}

```

2. Add an additional method called `positionSymbolsByKind`, which loops over `symbolsByKind` and calls `positionSymbols` for each one, like

```js
positionSymbolsByKind() {
  for (const kind of Object.keys(this.symbolsByViewport)) {
    const symbolsByViewport = this.symbolsByKind[kind];

    this.sketchPage.layers = [
      ...this.sketchPage.layers,
      ...this.positionSymbols(symbolsByViewport)
    ];
  }
}
```
3. Add an additional method called `positionSymbolsByViewport`:

```js
positionSymbolsByViewport() {
  this.sketchPage.layers = [
    ...this.sketchPage.layers,
    ...this.positionSymbols(this.symbolsByViewport)
  ];
}
```

hipstersmoothie replied:
Integrated this code ✅

chrisvxd commented on src/server/Story2sketch.js:
```diff
@@ -289,9 +327,12 @@ export default class Story2sketch {
+    if (this.byKind) {
+      this.writeByKind()
+    } else {
+      this.positionSymbols();
```
We should add a feature flag to toggle between `positionSymbolsByViewport()` and `positionSymbolsByKind()` - `layoutByKind`?

hipstersmoothie replied:
layoutByKind added ✅

hipstersmoothie commented on src/server/Story2sketch.js:
```diff
@@ -254,22 +259,55 @@ export default class Story2sketch {
+          }
         }
+
+        xOffset += this.widestByViewport[id] + this.symbolGutter;
```
code integrated ✅

hipstersmoothie commented on src/server/Story2sketch.js:
```diff
@@ -135,6 +139,12 @@ export default class Story2sketch {
     await pagePool.init();
 
     for (const { kind, stories } of this.stories) {
+      this.symbolsByViewport[kind] = {}
```
symbolsByKind implemented ✅

hipstersmoothie commented:
@chrisvxd I have added both `outputByKind` and `layoutByKind`. 

`layoutByKind` is a boolean that works exactly like your code above, I just had to add some code to account for the x position.

`outputByKind` I changed to be a path instead of a boolean. The way it is now when a user provides `outputByKind`: `out/directory`, story2sketch will create the individual files while still outputting a file with them all.

hipstersmoothie commented:
I think I’m gonna release a storybook plugin built in this pr. With this code I can add a toolbar icon to download the sketch files for the current displayed story kind 🤯

chrisvxd commented:
@hipstersmoothie you mean a fully blown story2sketch plugin? Awesome. That's been something I've thought about in the past.

Re: this PR - Haven't had a chance to look again (busy week) but will try to get around to it today. Would also appreciate your input to my comment in the original issue #63.

hipstersmoothie commented:
Yeah I'm thinking the tool could do a few things. For now it will just download the generated sketch file for the kind. 

It might be possible to run story2sketch on the current story rendering to produce a sketch file for whats on screen. I don't think that's possible now though

hipstersmoothie commented:
:neckbeard:

hipstersmoothie commented:
@chrisvxd :neckbeard:

chrisvxd commented:
@hipstersmoothie waiting on outcome of #63 before reviewing :)

chrisvxd commented:
Awaiting on API to be updated to match discussions in #63 before reviewing

hipstersmoothie commented:
@chrisvxd changed to match suggested API!

chrisvxd commented:
Thanks for the changes again @hipstersmoothie. Will review as soon as I can. Hopefully before I go away on Friday.

chrisvxd commented on src/server/Story2sketch.js:
```diff
@@ -251,26 +265,80 @@ export default class Story2sketch {
+      chalk.green(
+        `Success! ${
+          this.processedStories
+        } wrote stories by kind to ${chalk.white.bold(this.output)}`
```
Typo here, reads:

> Success! 3 wrote stories by kind to [path]

Should read

> Success! 3 stories written by kind to [path]

mick-feller commented:
Really hoping we can get this soon, i have a storybook with 168 stories and my json file is 25mb large. Sketch doesn't really like that.

Keep up the good work!

chrisvxd commented:
@mick-feller Yup shouldn't be long now. In the mean time, you could fork this branch and use that.

chrisvxd commented:
@hipstersmoothie going to address my most recent comments to get this in

chrisvxd commented:
Merged via #65! (PR comments addressed)

chrisvxd commented:
Released in v1.5.0
